mail_sendgrid/models/email_tracking.py

```diff
@@ -0,0 +1,169 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016-2017 Compassion CH (http://www.compassion.ch)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+import logging
+from datetime import datetime
+
+from werkzeug.useragents import UserAgent
+
+from odoo import models, fields, api
+
+_logger = logging.getLogger(__name__)
+
+
+class MailTrackingEmail(models.Model):
+    """ Count the user clicks on links inside e-mails sent.
+        Add tracking methods to process Sendgrid Notifications
+    """
+    _inherit = 'mail.tracking.email'
+
+    click_count = fields.Integer(compute='_compute_clicks', store=True)
+
+    @api.depends('tracking_event_ids')
+    def _compute_clicks(self):
+        for mail in self:
+            mail.click_count = len(mail.tracking_event_ids.filtered(
+                lambda event: event.event_type == 'click'))
+
+    @property
+    def _sendgrid_mandatory_fields(self):
+        return ('event', 'sg_event_id', 'timestamp',
+                'odoo_id', 'odoo_db')
+
+    @property
+    def _sendgrid_event_type_mapping(self):
+        return {
+            # Sendgrid event type: tracking event type
+            'bounce': 'hard_bounce',
+            'click': 'click',
+            'deferred': 'deferral',
+            'delivered': 'delivered',
+            'dropped': 'reject',
+            'group_unsubscribe': 'unsub',
+            'open': 'open',
+            'processed': 'sent',
+            'spamreport': 'spam',
+            'unsubscribe': 'unsub',
+        }
+
+    def _sendgrid_event_type_verify(self, event):
+        event = event or {}
+        sendgrid_event_type = event.get('event')
+        if sendgrid_event_type not in self._sendgrid_event_type_mapping:
+            _logger.error("Sendgrid: event type '%s' not supported",
+                          sendgrid_event_type)
+            return False
+        # OK, event type is valid
+        return True
+
+    def _db_verify(self, event):
```

Comment:
better to prefix this w/ `_sendgrid` as you did - nicely - w/ the others (unless this comes from the mail_tracking module ;P)